website/projects/aws/awssync_refactored.py

```diff
@@ -14,6 +19,86 @@ class AWSSyncRefactored:
+
+        for project in Project.objects.filter(mailinglist__isnull=False, semester=current_semester).values(
+            "slug", "semester", "mailinglist"
+        ):
+            project_slug = project["slug"]
+            project_semester = str(Semester.objects.get(pk=project["semester"]))
+            project_email = MailingList.objects.get(pk=project["mailinglist"]).email_address
+
+            sync_data = SyncData(project_email, project_slug, project_semester)
+            sync_data_list.append(sync_data)
+        return sync_data_list
+
+    def generate_aws_sync_list(self, giphouse_data: list[SyncData], aws_data: list[SyncData]) -> list[SyncData]:
+        """
+        Generate the list of users that are registered on the GiPhouse website, but are not yet invited for AWS.
+
+        This includes their ID and email address, to be able to put users in the correct AWS organization later.
+        """
+        return [x for x in giphouse_data if x not in aws_data]
+
+    def extract_aws_setup(self, parent_ou_id: str) -> AWSTree:
+        """
+        Give a list of all the children of the parent OU.
+
+        :param parent_ou_id: The ID of the parent OU.
+        :return: A AWSTree object containing all the children of the parent OU.
+        """
+        aws_tree = AWSTree(
+            "root",
+            parent_ou_id,
+            [
+                Iteration(
+                    ou["Name"],
+                    ou["Id"],
+                    member_accounts := [
+                        SyncData(
+                            account["Email"],
+                            next(
+                                (
+                                    d["Value"]
+                                    for d in self.api_talker.list_tags_for_resource(resource_id=account["Id"])
+                                    if d["Key"] == "project_slug"
+                                ),
+                                None,
+                            ),
+                            next(
+                                (
+                                    d["Value"]
+                                    for d in self.api_talker.list_tags_for_resource(resource_id=account["Id"])
+                                    if d["Key"] == "project_semester"
+                                ),
+                                None,
+                            ),
+                        )
+                    ],
+                )
+                for ou in self.api_talker.list_organizational_units_for_parent(parent_id=parent_ou_id)
+                for account in self.api_talker.list_accounts_for_parent(parent_id=ou["Id"])
+            ],
+        )
+        for member_account in member_accounts:
+            if not member_account.project_slug or not member_account.project_semester:
+                self.logger.warning(f"Account {member_account.project_email} has no project slug or semester tag.")
```

Comment:
should this really be a soft fail instead of a hard fail? What if somewhere else in the code we depend on having the tags?

Reply:
> should this really be a soft fail instead of a hard fail? What if somewhere else in the code we depend on having the tags?

I agree with this in hindsight (it being a soft fail is my fault). Make it throw an `Exception` that gets caught by pipeline and treated as a hard fail. It would be nicest if you include in the exception **all** accounts that have some `None` attribute. You can again do this with small list comprehension.

Also, you can apply De Morgan's law to the if-condition.

Reply:
BTW a hard fail can be triggered by any OU under root that has a member account without the appropriate tags. It should later be documented well that the pipeline assumes such a structure in order for the AWS synchronization process to work. 

An alternative would be to only check the current course OU for member accounts without the appropriate tags, instead of all OUs. Then such a structure is not strict anymore, and it makes sense to me because the entire synchronization process only applies to the current course OU (we filter projects on current semester, we create only current current course OU and do no synchronization with old ones, etc.).

What do you guys think? We have a few options here.

Reply:
I also propose the checking of the tags to be moved to to `Checks` class so this function only extracts the AWS tree as the function suggests.